topologymanager: Add Merge method to Policy

moshe010 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
In the current design the Topology Manager generates the merged
hints and select the best hint and the policy indicates whether
or not to admit the pod.

This patch adds Merge abstraction to Topology Manager policies which
allow the them to control on how to merge the permutation hints.
It also update the strict policy to admit only if all resources
aligned to same NUMA.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
This implements the change proposed in https://github.com/kubernetes/enhancements/pull/1160
Only the non-merge commits are part of the actual PR, 
This PR is depened on https://github.com/kubernetes/kubernetes/pull/80569 and https://github.com/kubernetes/kubernetes/pull/80570

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

moshe010 commented:
/cc @lmdaly @klueska

moshe010 commented:
/cc @ConnorDoyle

klueska commented:
/ok-to-test

klueska commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -166,20 +176,14 @@ func (m *manager) calculateAffinity(pod v1.Pod, container v1.Container) Topology
-				klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity with resource '%s'", resource)
-				affinity, _ := socketmask.NewSocketMask()
-				affinity.Fill()
-				allProviderHints = append(allProviderHints, []TopologyHint{{affinity, true}})
```
Removing this from here and above changes the semantics of the existing code slightly.
In both of these cases we are essentially creating a `defaultHint` of {full, true} (which will override the original default hint of {full, false} when hints are merged.

With your changes, the result of all resource types returning a `nil` will not result in {full, true} (as it should), but rather {full, false} because the default will never be overridden.

moshe010 replied:
In case I will put it back then in policy strict it will mean that we will failed to admit the pod (in case of one resource not having hints) . if that is the required behavior then I would put it back.  So I mean that the for best-effort it will be don't care and for strict it will mean don't admit.

klueska replied:
Yeah, we had this discussion before (somewhere other than this PR it seems, but I don't remember where). We concluded that the semantics encoded here are highly dependent on the `best-effort` policy.

As such, I actually pictured the abstraction you are building here to allow customization of the entire `calculateAffinity()` call and not just a replacement of the merge strategy.

moshe010 replied:
yes it was in the resource map. The reason I did it like this was that for best-effort I don't think it matter if the hint is {full, ture} or {full, false} (please correct me if I am wrong), but it does affect the strict policy. 
Customization of the calculateAffinity is a bit more complicated and unfortunately I on PTO until end of month.   I think the question is if this is good enough for current release. 
Anyway  adrianchiris "adrianc@mellanox.com" from my team will continue working this if required.

adrianchiris replied:
strict policy should not prevent pod admission in case a resource does not care about affinity.
on the bright side, SR-IOV (network) and CPU resources do care about affinity.

Anyway, I still need to properly catch up with related PR, will try to do that today.

klueska replied:
I’d prefer it if the new Merge() call actually encapsulated what right now is called calculateAffinity(). It might lead to some amount of duplicated code, but my guess is that it will be less than we expect if we clean up the strict implementation a bit.

The current implementation of calculateAffinity() will then simply become the new Merge() call for the best effort policy, and a new (simplified) one can be written for strict.

adrianchiris replied:
makes sense, i believe default hint generation should be per policy,
(or allow a policy to generate a default hint)
i will work on it next week.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_strict.go:
```diff
@@ -69,3 +72,97 @@ func (p *strictPolicy) Merge(permutation []TopologyHint) TopologyHint {
+	// Loop through all provider hints and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allProviderHints [][]TopologyHint
```
Now that policy is in charge of selecting the best hint we can probably improve this by only appending hints with a single bit set in the socketmask as any permutation containing a hint with more than one bit set will lead to preferred=false in the selected bestHint.

moshe010 commented:
It hard to resolve the conflict on this PR. (too many changes merged to master) 
I am currently working on another PR that will incorporate the changes from here. Once it will be pushed I will close this one.

adrianchiris committed:
Extend Policy with Merge method

This commit performs some refactoring of topology manager
it moves the responsibility of merging topology hints to the Policy
object.

In addition, for single numa node policy, merge logic has been
split off restricted/best-effort and a new method for merging
topology hints is introduced.

The following improvements were introduced:

- Change the current method of finding the best hint.
  Instead of going over all permutations, sort the hints and find
  the narrowest hint common to all resources.
- Break out early in when merging to a preferred hint is not possible

Unit tests have been updated to reflect the changes made.

adrianchiris commented:
/retest pull-kubernetes-integration

adrianchiris commented:
/assign @klueska

adrianchiris commented:
/cc @lmdaly

klueska commented on pkg/kubelet/cm/topologymanager/policy.go:
```diff
@@ -20,10 +20,12 @@ import (
+	// Returns Pod Admit Handler Response based on hints and policy type
 	CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAdmitResult
+	// Returns a merged TopologyHint based on input from hint providers
+	Merge(providersHints []map[string][]TopologyHint, numaNodes []int) TopologyHint
```
Instead of passing numaNodes here, can we have the policy take `cputopology.NUMANodeInfo` as a parameter when instantiated and store the `numaNodes` then?

adrianchiris replied:
sure

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -48,3 +51,134 @@ func (p *singleNumaNodePolicy) CanAdmitPodResult(hint *TopologyHint) lifecycle.P
+	return filteredResourcesHints
+}
+
+func (p *singleNumaNodePolicy) Merge(providersHints []map[string][]TopologyHint, numaNodes []int) TopologyHint {
```
The implementation for `Merge()` on the single-numa-node plicy seems much more complicated than I would expect. I haven't looked at this code in detail yet, but can you explain in words what it is doing and why it needs to be as complicated as it is?

adrianchiris replied:
In high level, we do the following:

1. create a list of all resources hints (`allResourcesHints [][]TopologyHint`) from providersHints
    1.1 Break out early if we encounter a resource with no possible socket affinities.
     1.2 Resources/providers with no preference for socket affinity dont end up in the final list
2. out of `allResourcesHints` filter out any hints with !=1 bits set in the bit mask or hints with nil affinity
3. check that the filtered list is not empty and does not have an empty element, i.e all required
    resources have at least one hint. If this is not the case, break out early.
4. Get the best hint
    4.1 sort hints in every resource
    4.2 the best hint is the narrowest hint found in all resources.

klueska commented on pkg/kubelet/cm/topologymanager/topology_manager_test.go:
```diff
@@ -104,22 +104,18 @@ func TestGetAffinity(t *testing.T) {
 	}
 }
 
-func TestCalculateAffinity(t *testing.T) {
```
It's hard to verify taht all the same tests are still here after this change. Can you walk me through your changes a bit. It looks like there are some additions, some removals, and some outright changes to the tests.

Ideally this would have been split into 3 separate commits to make it easier to review (one for changes, one for additions, and one for removals). But if you can walk me through the changes a bit more that should be OK:

adrianchiris replied:
So 90% of the tests where moved to the policies.
the main change is that the merge method now accepts : `[]map[string][]TopologyHint`  
so the test case input needed to be changed slightly as it no longer deals with hint providers directly.

in topology manager we test that it is able to properly coalesce the info from hint providers and pass it to the policy.Merge method.

The tests surrounding merge where moved to `policy_best_effort_test.go:testOptimalMerge`
with the needed change of translating hint providers  `GetTopologyHints` output to `[]map[string][]TopologyHint`

The test cases that where removed is because they where not valid for the new Mege method.
however what those tests were testing was kept in `TestCalculateAffinity` (manly around edge cases of what a hint provider may return in `GetTopologyHints`)

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -41,3 +43,144 @@ func (p *bestEffortPolicy) CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAd
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with resource '%s'", resource)
+				allProviderHints = append(allProviderHints, []TopologyHint{{defaultAffinity, true}})
```
Do we need the fix now that the Single NUMA node policy has a separate Merge() logic ?

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -41,3 +43,144 @@ func (p *bestEffortPolicy) CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAd
+	// found that has at least one NUMA ID set, return the 'defaultAffinity'.
+	bestHint := TopologyHint{defaultAffinity, false}
+	p.iterateAllProviderTopologyHints(allProviderHints, func(permutation []TopologyHint) {
+
```
will remove neline

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -41,3 +43,144 @@ func (p *bestEffortPolicy) CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAd
+		}
+	}
+	iterate(0, []TopologyHint{})
+}
```
yes it is unchanged, it was merely moved from `topology_manager.go`

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_restricted.go:
```diff
@@ -20,7 +20,9 @@ import (
-type restrictedPolicy struct{}
+type restrictedPolicy struct {
+	bestEffortPolicy
+}
```
Yes,
later we can consider introducing some improvement to the way hints are merged for restricted policy.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy.go:
```diff
@@ -20,10 +20,12 @@ import (
 	Name() string
-	//Returns Pod Admit Handler Response based on hints and policy type
+	// Returns Pod Admit Handler Response based on hints and policy type
 	CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAdmitResult
```
I wanted to reuse the Merge logic from best effort policy in restricted.
TBH i dont see a major added value in having the two combined. can you elaborate on why you feel this is needed ?

klueska replied:
I don't feel that strongly about it, it's just how I pictured it working once we had the `Merge()` abstraction in place.

I was picturing a helper function in `policy.go` that  implemented the bit-wise AND logic, that both the `best-effort` and `restricted` policies called out to. Their respective `Merge()` implementations then inspected the bit-wise ANDed `TopologyHint` to decide whether to return `true` or `false` for admit.

I'm fine leaving it the way it is so long as it's clean and easy to understand.

klueska replied:
It just feels weird (as an abstraction) to have one call in the abstraction return a result, that is then immediately passed to another call in the same abstraction in all cases.

adrianchiris replied:
Thinking about this further, I tend to agree about the weirdness here.
Will simplify the interface by removing `CanAdmitPodResult` and adjust unit tests.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_restricted_test.go:
```diff
@@ -56,3 +74,7 @@ func TestPolicyRestrictedCanAdmitPodResult(t *testing.T) {
+
+func TestPolicyRestrictedMerge(t *testing.T) {
+	testOptimalMerge(t, NewRestrictedPolicy())
+}
```
The one difference, today, between restricted and best-effort is reflected in TestPolicyRestrictedCanAdmitPodResult.  (not counting the name of the policy :))

this ensures that Restricted policy merge method behaves the same as best effort policy merge method.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort_test.go:
```diff
@@ -47,3 +65,408 @@ func TestPolicyBestEffortCanAdmitPodResult(t *testing.T) {
+	testOptimalMerge(t, NewBestEffortPolicy())
+}
+
+func testOptimalMerge(t *testing.T, policy Policy) {
```
Today, the same Merge() is used in Best effort and Restricted policies.
having the decision whether or not to admit a pod split from the merge makes this possible.

adrianchiris commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -94,6 +94,26 @@ type TopologyHint struct {
+	}
+	return false
+}
+
```
while not directly related to what we aim to achieve here, but i guess why not since we are already refactoring...

adrianchiris commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -94,6 +94,26 @@ type TopologyHint struct {
+	nodeAffinity := "nil"
+	if th.NUMANodeAffinity != nil {
+		nodeAffinity = th.NUMANodeAffinity.String()
+	}
```
will remove.

adrianchiris committed:
Address comments after code review

- Remove redundant String() method from TopologyHint
- Add LessThan() method to TopologyHint
- Move numaNodes parameter from Merge() to an attribute
  in the policy object
- Merge policy's CanAdmitPodResults() into Merge() for simplicity
  and cleaness
- Adjust unit tests to reflect the changes

adrianchiris commented:
/test pull-kubernetes-verify

nolancon commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -105,20 +103,29 @@ func (th *TopologyHint) IsEqual(topologyHint TopologyHint) bool {
+		return th.Preferred == true
 	}
-	return fmt.Sprintf("(Preferred=%t, NUMANodeAffinity=%s)", th.Preferred, nodeAffinity)
+	return th.NUMANodeAffinity.IsNarrowerThan(b.NUMANodeAffinity)
```
'b.NUMANodeAffinity' should be 'other.NUMANodeAffinity' ?

adrianchiris replied:
Yes! will sort it out

bg-chun commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -17,27 +17,33 @@ limitations under the License.
-func (p *singleNumaNodePolicy) CanAdmitPodResult(hint *TopologyHint) lifecycle.PodAdmitResult {
-	if !hint.Preferred || hint.NUMANodeAffinity.Count() > 1 {
+func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.PodAdmitResult {
+	if !hint.Preferred || hint.NUMANodeAffinity.Count() != 1 {
```
Do we have to consider whether the final hint is preferred or not?
I don't know how the device-plugin generates hint and set the preferred flag.

adrianchiris replied:
since the logic is now separate, it is possible rely on either IMO.

please note that the changes (apart of the merge logic change) is already merged in PR #84721

This PR needs to be modified to propose the merge logic change only.
(or alternatively kill this PR and propose a new one)

adrianchiris commented:
A rebased version of this PR containing the single numa node policy modifications is proposed in PR#85798.

klueska commented:
/close superceded by #85798

klueska commented:
/close
superceded by #85798
